server/src/main/java/io/spine/server/storage/memory/InMemoryInboxStorage.java

```diff
@@ -77,38 +78,62 @@ TenantInboxRecords createSlice() {
     }
 
     @Override
-    public void write(InboxMessage message) {
+    public synchronized Optional<InboxMessage> oldestMessageToDeliver(ShardIndex index) {
+        TenantInboxRecords storage = multitenantStorage.currentSlice();
+        Optional<InboxMessage> result =
+                storage.readAll()
+                       .stream()
+                       .filter((r) -> index.equals(r.getShardIndex()) &&
+                               r.getStatus() == InboxMessageStatus.TO_DELIVER)
+                       .min(InMemoryInboxStorage::compareMessages);
+        return result;
+    }
+
+    private static int compareMessages(InboxMessage m1, InboxMessage m2) {
+        int timeComparison = compare(m1.getWhenReceived(), m2.getWhenReceived());
+        if (timeComparison != 0) {
+            return timeComparison;
+        }
+        int versionComparison = Integer.compare(m1.getVersion(), m2.getVersion());
+        if(versionComparison == 0) {
+            throw new IllegalStateException("Versions must not be equal.");
+        }
+        return versionComparison;
+    }
+
+    @Override
+    public synchronized void write(InboxMessage message) {
         multitenantStorage.currentSlice()
                           .put(message.getId(), message);
     }
 
     @Override
-    public void writeAll(Iterable<InboxMessage> messages) {
+    public synchronized void writeAll(Iterable<InboxMessage> messages) {
```

Comment:
Please document why are the methods `synchronized`.

Reply:
All of the mutating methods were `synchronized` already except for this one. The corresponding explanation added at the top of the class.